crates/subspace-farmer/src/single_disk_farm/farming.rs

```diff
@@ -146,57 +152,89 @@ where
-        });
-
-        for (sector_index, solution_candidates) in solution_candidates {
-            for maybe_solution in solution_candidates.into_iter::<_, PosTable>(
-                &reward_address,
-                &kzg,
-                &erasure_coding,
-                &mut table_generator,
-            )? {
-                let solution = match maybe_solution {
-                    Ok(solution) => solution,
-                    Err(error) => {
-                        error!(%slot, %sector_index, %error, "Failed to prove");
-                        // Do not error completely on disk corruption or other
-                        // reasons why proving might fail
-                        continue;
+                .into_iter()
+                .filter_map(|(sector_index, solution_candidates)| {
+                    let sector_solutions = match solution_candidates.into_solutions(
+                        &reward_address,
+                        &kzg,
+                        &erasure_coding,
+                        |seed: &PosSeed| table_generator.lock().generate_parallel(seed),
+                    ) {
+                        Ok(solutions) => solutions,
+                        Err(error) => {
+                            warn!(
+                                %error,
+                                %sector_index,
+                                "Failed to turn solution candidates into solutions",
+                            );
+
+                            return None;
+                        }
+                    };
+
+                    if sector_solutions.len() == 0 {
+                        return None;
                     }
-                };
 
-                debug!(%slot, %sector_index, "Solution found");
-                trace!(?solution, "Solution found");
-
-                solutions.push(solution);
+                    Some((sector_index, sector_solutions))
+                })
+                .collect::<Vec<_>>();
+
+            sectors_solutions.sort_by(|a, b| {
+                let a_solution_distance =
+                    a.1.best_solution_distance().unwrap_or(SolutionRange::MAX);
+                let b_solution_distance =
+                    b.1.best_solution_distance().unwrap_or(SolutionRange::MAX);
+
+                // Comparing `b` to `a` because we want smaller values first
+                b_solution_distance.cmp(&a_solution_distance)
+            });
+
+            let mut solutions = Vec::<Solution<PublicKey, PublicKey>>::new();
+
+            for (sector_index, sector_solutions) in sectors_solutions {
+                for maybe_solution in sector_solutions {
+                    let solution = match maybe_solution {
+                        Ok(solution) => solution,
+                        Err(error) => {
+                            error!(%slot, %sector_index, %error, "Failed to prove");
+                            // Do not error completely on disk corruption or other
```

Comment:
Could be `Do not error completely as disk corruption or other reasons why proving might fail`